app/javascript/mastodon/actions/compose.js

```diff
@@ -390,13 +391,13 @@ export function hydrateCompose() {
   };
 }
 
-function insertIntoTagHistory(tags) {
+function insertIntoTagHistory(text) {
   return (dispatch, getState) => {
     const state = getState();
     const oldHistory = state.getIn(['compose', 'tagHistory']);
     const me = state.getIn(['meta', 'me']);
-    const names = tags.map(({ name }) => name);
-    const intersectedOldHistory = oldHistory.filter(name => !names.includes(name));
+    const names = text.match(APPROX_HASHTAG_RE).map(name => name.replace(/^\s?#/, ''));
```

Comment:
Instead of `APPROX_HASHTAG_RE`, you'd use something like `new RegExp('(?:^|[^\\/\)\\w])(' + response.data.tags.map(escapeRegExp).join('|') + ')', 'i')`, I guess.

See 95c8232109f5e59273de48858ede7fbe7ce63d58 for why the regexps don't match between ruby and javascript